### PR TITLE
Add tests for dict module

### PR DESCRIPTION
**Pull request by fmoletta**
TESTING STATUS:

**Dict Module**

- `dict_new():` PASSING 

- `dict_read()`: FAILING (Issue with ids structs)

- `dict_write()`: FAILING (Issue with ids structs)

- `dict_update()`: FAILING (Issue with ids structs)

- `default_dict_new():` PASSING 

- `squash_dict()`: FAILING (Issue with ids structs)

- `dict_squash()`: FAILING (Issue with ids structs)

### CODE REVIEW

File: cairo_programs/default_dict_new.cairo [hunk 1]
```diff
@@ -0,0 +1,19 @@
+from starkware.cairo.common.dict_access import DictAccess
+
+# Creates a new dictionary, with a default value.
+func default_dict_new(default_value : felt) -> (res : DictAccess*):
+    %{
```
Comment (pefontana) [inline cairo_programs/default_dict_new.cairo:5]: Should we add the # TEST tag?

File: cairo_programs/default_dict_new.cairo [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+            __dict_manager = DictManager()
+
+        memory[ap] = __dict_manager.new_default_dict(segments, ids.default_value)
+    %}
+    ap += 1
+    return (res=cast([ap - 1], DictAccess*))
+end
+
+func main():
+    let my_dict = default_dict_new(17)
```
Comment (pefontana) [inline cairo_programs/default_dict_new.cairo:17]: It is possible to add a simple assert here?

File: cairo_programs/dict_new.cairo [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        memory[ap] = __dict_manager.new_dict(segments, initial_dict)
+        del initial_dict
+    %}
+    ap += 1
+    return (res=cast([ap - 1], DictAccess*))
+end
+
+func main():
+    %{ initial_dict = {1:2, 2:3, 4:5}%}
+    let my_dict = dict_new()
```
Comment (pefontana) [inline cairo_programs/dict_new.cairo:22]: Same here, it is possible to add a simple assert here?

**Review by pefontana — CHANGES_REQUESTED** [review]

File: cairo_programs/default_dict_new.cairo (hunk 1 above)
Comment (fmoletta) [reply cairo_programs/default_dict_new.cairo:5]: Fixed [6c2623b]

File: cairo_programs/default_dict_new.cairo (hunk 2 above)
Comment (fmoletta) [reply cairo_programs/default_dict_new.cairo:18]: Not really, the only thing that changes in cairo is that an address gets inserted into ap, and we can't assert that

File: cairo_programs/dict_new.cairo (hunk 3 above)
Comment (fmoletta) [reply cairo_programs/dict_new.cairo:22]: Same case

File: cairo_programs/dict_read.cairo [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+end
+
+func main():
+    alloc_locals
+    %{ initial_dict = {1:2, 2:3, 4:5}%}
+    let (my_dict) = dict_new()
+    let (local val1 : felt) = dict_read{dict_ptr=my_dict}(key=1)
+    assert val1 = 2
+    let (local val2 : felt) = dict_read{dict_ptr=my_dict}(key=2)
+    assert val1 = 3
```
Comment (pefontana) [inline cairo_programs/dict_read.cairo:27]:
```suggestion
    assert val2 = 3
```

File: cairo_programs/dict_read.cairo [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    let (local val3 : felt) = dict_read{dict_ptr=my_dict}(key=4)
+    assert val1 = 5
```
Comment (pefontana) [inline cairo_programs/dict_read.cairo:29]:
```suggestion
    assert val3 = 5
```

File: cairo_programs/dict_update.cairo [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    return ()
+end
+
+func main():
+    alloc_locals
+    %{ initial_dict = {1:2, 2:3, 4:5}%}
+    let (my_dict) = dict_new()
+    dict_update{dict_ptr=my_dict}(key=1, prev_value=2, new_value=5)
+    let (local val : felt) = dict_read{dict_ptr=my_dict}(key=1)
+    assert val = 34
```
Comment (pefontana) [inline cairo_programs/dict_update.cairo:35]:
```suggestion
```

**Review by pefontana — CHANGES_REQUESTED** [review]
We will be good to go with these minor corrections in the cairo programs.
If you give me the OK I can commit them, if you want.
Also the dict_red, dict_write and dict update tests continue failing doe to `Custom Hint Error: AttributeError: 'PyTypeId' object has no attribute 'segment_index'`

**Review by Oppen — APPROVED** [review]

**Review by Juan-M-V — APPROVED** [review]

**Comment by fmoletta** [comment]
> We will be good to go with these minor corrections in the cairo programs.
> If you give me the OK I can commit them, if you want.
> Also the dict_red, dict_write and dict update tests continue failing doe to `Custom Hint Error: AttributeError: 'PyTypeId' object has no attribute 'segment_index'`

Committed and checked the changes.
My suspicion is that PyTypeId__getattr__ is failing to properly construct a PyMaybeRelocatable, as the error arrises from trying to access segment_index as a struct field, rather than a PyRelocatable attribute (expected behavior).

**Pull request merged**